kong.conf.default

```diff
@@ -110,6 +110,11 @@
                                  # `m`, with a minimum recommended value of
                                  # a few MBs.
 
+#proxy = on                      # Determines if Nginx should be listening for
```

Comment:
If this is `off`, then `ssl = on` should probably have no impact whether the proxy server is created or not. In other words, this `proxy` setting should be considered top-level, and take precedence over both `proxy_listen` and `admin_listen`. Ditto for `admin`. Thoughts?

Reply:
Yes, I expected that comment, this was to be discussed. The issue is with the naming, because it is called `proxy` which tends to lead to believe the entire proxy part is disabled, where it only disables the `http` port now.

What I implemented was basically 4 settings (2 existing, 2 new), controlling all 4 ports: proxy, proxy-ssl, admin and admin-ssl. This allows for maximum flexibility. eg. a configuration with only `proxy-ssl` enabled would not be possible if `proxy` were `top-level` as you suggested.

So this current implementation is not clear. Imo we have 4 options:

1. Find a better name for `proxy` (considering that `ssl` is its counterpart), and `admin` (considering `admin_ssl` is that ones counterpart)
2. Provide less flexibility, where `admin` and `proxy` are both "top-level" and control both their respective http and https ports
3. rename all 4 parameters, such that the names make sense. But this would probably be a breaking change when the existing 2 are renamed.
4. over-the-top: do 3, but also add 2 'top-level' ones.

I like the flexibility, so I'd go with 1 or 3.

Reply:
I have given this quite a lot of thought in the past already, and my belief is that some configuration values are cases of "over-configuration", and could be factored together. Some other configuration properties could be set to sane defaults too.

Something like:
```
proxy_listen = off | 127.0.0.1:80 | 127.0.0.1:443 ssl
```

(the values are comma separated).

This way:
```
# control plane
proxy_listen = off

# data plane
proxy_listen = 127.0.0.1:80, 127.0.0.1:443 ssl
admin_listen = off

# https only, both planes
proxy_listen = 127.0.0.1:443 ssl
admin_listen = 127.0.0.1:8443 ssl
```

Reply:
That would introduce a completely new way of setting options, more complex. Try explaining that in the template conf file, in 2 lines of text. Not saying we shouldn't, just that it might turn into a can of worms.

Also might give the impression that you could do:

> proxy_listen = 127.0.0.1:80, 127.0.0.1:81, 127.0.0.1:443 ssl, 127.0.0.1:444 ssl

Reply:
I think of it as a much more flexible way to configure Kong (and simpler too). Yes, this does mean we can listen on several interfaces/ports (awesome). This means one can now do:
```
proxy_listen = 127.0.0.1:80, [::1]:80
``` 

What is there not to want about this, really? (this is rhetorical :D)

Reply:
> That would introduce a completely new way of setting options, more complex.

Let's be honest: we need new ways of doing a lot of things that we have been (wrongly) assuming for so long now... Time to break away from a few hastily designed patterns and move forward.

Reply:
> and could be factored together

A hundred +1.